apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,244 @@
+	// FederationDomain to further customize how these usernames are presented to Kubernetes.
+	//
+	// Defaults to "login:id", which is the user login attribute, followed by a colon, followed by the unique and
+	// unchanging integer ID number attribute. This blends human-readable login names with the unchanging ID value
+	// from GitHub. Colons are not allowed in GitHub login attributes or ID numbers, so this is a reasonable
+	// choice to concatenate the two values.
+	//
+	// See the response schema for
+	// [Get the authenticated user](https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user).
+	//
+	// +kubebuilder:default="login:id"
+	// +kubebuilder:validation:Enum={"id","login","login:id"}
+	// +optional
+	Username *GitHubUsernameAttribute `json:"username"`
+
+	// Groups configures which property of the GitHub team record shall determine the group names in Kubernetes.
+	//
+	// Can be either "name" or "slug". Defaults to "slug".
+	//
+	// GitHub team names can contain upper and lower case characters, whitespace, and punctuation (e.g. "Kube admins!").
+	//
+	// GitHub team slugs are lower case alphanumeric characters and may contain dashes and underscores (e.g. "kube-admins").
+	//
+	// Group names as presented to Kubernetes will always be prefixed by the GitHub organization name followed by a
+	// forward slash (e.g. "my-org/my-team"). GitHub organization login names can only contain alphanumeric characters
+	// or single hyphens, so the first forward slash `/` will be the separator between the organization login name and
+	// the team name or slug.
+	//
+	// If desired, an admin could configure identity transformation expressions on the Pinniped Supervisor's
+	// FederationDomain to further customize how these group names are presented to Kubernetes.
+	//
+	// See the response schema for
+	// [List teams for the authenticated user](https://docs.github.com/en/rest/teams/teams?apiVersion=2022-11-28#list-teams-for-the-authenticated-user).
+	//
+	// +kubebuilder:default=slug
+	// +kubebuilder:validation:Enum=name;slug
+	// +optional
+	Groups *GitHubGroupNameAttribute `json:"groups"`
+}
+
+// GitHubClientSpec contains information about the GitHub client that this identity provider will use
+// for web-based login flows.
+type GitHubClientSpec struct {
+	// SecretName contains the name of a namespace-local Secret object that provides the clientID and
+	// clientSecret for an GitHub App or GitHub OAuth2 client.
+	//
+	// This secret must be of type "secrets.pinniped.dev/github-client" with keys "clientID" and "clientSecret".
+	//
+	// +kubebuilder:validation:MinLength=1
+	SecretName string `json:"secretName"`
+}
+
+// GitHubIdentityProviderSpec is the spec for configuring an GitHub identity provider.
+type GitHubIdentityProviderSpec struct {
+	// GitHubApi allows configuration for GitHub Enterprise Server
+	//
+	// +kubebuilder:default={}
+	GitHubApi GitHubAPIConfig `json:"gitHubAPI,omitempty"`
```

Comment:
For the `json: `... Quick proper name question.  Should it be:

```yaml
gitHubAPI # camel, with first lowercase
githubAPI  # all lowercase, including "h"
```
The branded name is `GitHub`, but it is one word, not two, so there is a little confusion.  So should we treat it like "github" as one word, or "git hub" as two words?

Reply:
The Golang style guide from Google does not explicitly call out the rules for camel casing. The closest thing that it explicitly says is how to deal with acronyms inside variable names (see https://google.github.io/styleguide/go/decisions#initialisms.) Without the style guide directly advising us, we just have to pick one and be internally consistent. The guide says that DDoS should be used as `ddosVarName` or `varNameDDoS` for unexported variables, but should be `DDoSVarName` for exported variables. The guidance is to downcase all the letters when the word is the first word in an unexported variable, but keep the original case of each letter when the word is the first word in an exported variable. If we use that as an analogy then that would imply `GitHubAPI` for an exported variable and `githubAPI` for an unexported variable, right?

Not that we have to make the same choice, but as a point of reference Dex picked lower-casing the word when used in unexported variable names, e.g. `githubConnector` (see https://github.com/dexidp/dex/blob/v2.39.0/connector/github/github.go#L78).

So maybe this line should be:
```golang
GitHubApi GitHubAPIConfig `json:"githubAPI,omitempty"`
```
?

Regardless of how we name variables and JSON keys, we should be careful to always use `GitHub` in comments and documentation, because that it is name of the company.

Reply:
Another reference which does not say exactly what to do in this case: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions

Reply:
For this case I do like `GitHubAPI GitHubAPIConfig `json:"githubAPI,omitempty"``